drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,820 @@
+	    (data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			      uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+	uint32_t dev_config;
+	uint8_t *internal_buf = config->merge_buf;
+	uint16_t merge_buf_size = config->merge_buf_size;
+	uint16_t current_internal_buf_size = 0;
+	uint8_t *transaction_buf;
+	uint16_t transaction_len;
+	struct i2c_msg transaction_msg;
+	int ret = 0;
+
+	if (data->is_target) {
+		/* Currently target is registered, initiating transfer is not allowed */
+		return -EBUSY;
+	}
+
+	/* Transmit each message */
+	for (int i = 0; i < num_msgs; i++) {
+		bool merge_next_msg =
+			(i + 1 < num_msgs) && !(msgs[i].flags & I2C_MSG_STOP) &&
+			!(msgs[i + 1].flags & I2C_MSG_RESTART) &&
+			((msgs[i].flags & I2C_MSG_WRITE) == (msgs[i + 1].flags & I2C_MSG_WRITE));
+
+		if (merge_next_msg || current_internal_buf_size != 0) {
+			if ((current_internal_buf_size + msgs[i].len) > merge_buf_size) {
+				LOG_ERR("Need to use the internal driver "
+					"buffer but its size is insufficient "
+					"(%u + %u > %u). ",
+					current_internal_buf_size, msgs[i].len, merge_buf_size);
+				ret = -ENOSPC;
+				break;
+			}
+			if (!(msgs[i].flags & I2C_MSG_READ)) {
+				memcpy(internal_buf + current_internal_buf_size, msgs[i].buf,
+				       msgs[i].len);
+			}
+			current_internal_buf_size += msgs[i].len;
+		}
+
+		/* Continue merging messages */
+		if (merge_next_msg) {
+			continue;
+		}
+
+		/* No merge was performed, send standard frame */
+		if (current_internal_buf_size == 0) {
+			transaction_buf = msgs[i].buf;
+			transaction_len = msgs[i].len;
+		} else {
+			transaction_buf = internal_buf;
+			transaction_len = current_internal_buf_size;
+		}
+		/* Fill transaction_msg */
+		transaction_msg.flags = msgs[i].flags;
+		transaction_msg.buf = transaction_buf;
+		transaction_msg.len = transaction_len;
+
+		k_sem_take(data->i2c_busy_sem, K_FOREVER);
+
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+			ret = i2c_mspm0_receive(dev, transaction_msg, addr);
+		} else {
+			ret = i2c_mspm0_transmit(dev, transaction_msg, addr);
+		}
+
+		if (ret != 0) {
+			break;
+		}
+
+		/* For merged reads, populate data in original user buffers */
+		if ((transaction_msg.flags & I2C_MSG_READ) && (transaction_buf == internal_buf)) {
+			int j = i;
+
+			while (j >= 0 && current_internal_buf_size >= msgs[j].len) {
+				current_internal_buf_size -= msgs[j].len;
+				memcpy(msgs[j].buf, transaction_buf + current_internal_buf_size,
+				       msgs[j].len);
+				j--;
+			}
+		}
+		/* Safe to reset internal buffer size back to 0 at this point */
+		current_internal_buf_size = 0;
+	}
+
+	if (ret == -ETIMEDOUT) {
+		i2c_mspm0_get_config(dev, &dev_config);
+		i2c_mspm0_reset_peripheral_controller(dev);
+	}
+
+	k_sem_give(data->i2c_busy_sem);
```

Comment:
release after configure in case of failure? that will eventually keep all the timeout handling in one place